tools/builder/api.rb

```diff
@@ -0,0 +1,263 @@
+# Copyright 2019 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'net/http'
+require 'json'
+require 'active_support/inflector'
+
+require 'api/product'
+require 'api/resource'
+require 'api/type'
+require 'api/compiler'
+require 'api/async'
+
+# All properties are read-only in Magic Modules
+# Creating an api.yaml involves a lot of setting values.
+# This will create setters for all fields on an api.yaml
+# (but only in the context of the linter)
+# rubocop:disable Style/MissingRespondToMissing
+module Api
+  # Api::Object class being overriden.
+  class Object
+    # Create a setter if the setter doesn't exist
+    # Yes, this isn't pretty and I apologize
+    def method_missing(method_name, *args)
+      matches = /([a-z_]*)=/.match(method_name)
+      super unless matches
+      create_setter(matches[1])
+      method(method_name.to_sym).call(*args)
+    end
+
+    def create_setter(variable)
+      self.class.define_method("#{variable}=") { |val| instance_variable_set("@#{variable}", val) }
+    end
+  end
+end
+# rubocop:enable Style/MissingRespondToMissing
+
+TYPES = {
+  'string': 'String',
+  'boolean': 'Boolean',
+  'object': 'Map',
+  'integer': 'Integer',
+  'number': 'Double',
+  'array': 'Array'
+}.freeze
+
+# Handles all logic of merging a discovery + handwritten api.yaml in proper order.
+class HumanApi
```

Comment:
Seems strange to have a utility class, that just merges 2 objects. I would turn this into a functional method rather than encapsulate into a class with a `build` method.